### DISCUSSION TITLE
You need to enable JavaScript to run this app

### DISCUSSION THREAD

**Discussion by cata008** [Azure/ipam#126] [General]
@DCMattyG @hbendana 
I am trying to configure IPAM by using Azure Container Apps. While trying to connect to the UI, I end up in a loop and I am getting the error below: 
(api/users/me?expand=true)
**You need to enable JavaScript to run this app**

Any idea how I could solve this issue? 

Note: Java script is allowed in my browser (Chrome/Edge)

**Comment by DCMattyG** [comment]
Hey @cata008, please help me with a bit of clarity on this....

Are you actually navigating to ```https://<YourIPAMAddress>/api/users/me?expand=true``` in your web browser? If so, this is an API endpoint and needs an authorization header sent along with it using a program such as Postman.

Do you see the same error if you navigate to ```https://<YouIPAMAddress>/```?

I'm so sorry if I didn't understand exactly what is happening. If I'm missing something, perhaps a screenshot might help me have a bit more clarity on what you are facing.

**Reply by cata008** [reply]
@DCMattyG, thanks for your quick response!
I am navigating to my frontend container app, e.g. https://{frontend}.westeurope.azurecontainerapps.io/, I am successfully authenticating with my credentials against Azure AD, but then it remains stuck - see image below.

<img width="659" alt="image" src="https://user-images.githubusercontent.com/37331541/236847701-16b65b3a-31f5-4f5f-ade4-b9c55718cc1c.png">

**Reply by DCMattyG** [reply]
Interesting. I've never run the Azure IPAM solution in a Container App. Are you able to provide any details about your deployment/configuration so I can test this on my end?

We've only ever run this in an Azure App Service (or Function App if using the API only), and we've had some other customer us AKS, but ACA is a  new one for me.

Happy to help you troubleshoot, just need a few more details here 😉

**Reply by DCMattyG** [reply]
Also, that message in the *Preview* tab is the default message from the React index.html page that would show in the browser pane if Javascript isn't enabled for some reason (instead of the spinning circle you are currently seeing).

**Reply by cata008** [reply]
I noticed that the backend was not actually known to the frontend - so I added the additonal environment variable - REACT_APP_IPAM_ENGINE_URL.

However, since the backend uses a different URL, when I log on the UI, I am now getting a CORS error:
<img width="518" alt="image" src="https://user-images.githubusercontent.com/37331541/236875946-9a65a361-a897-46d6-bc6a-7dcc034861e9.png">

Regarding the deployment, I have basically deployed a container app environment that comprises the frontend and backend container apps.

I have tested the backend with Postman and it seems to work (apart from the reservation API which that returns an internal server error, but I'll open a new discussion if needed).

At this point, I deployed a public ingress for each container apps to make testing easier but that will change in the future.


Any idea how we could fix the CORS issue? 

I think the backend needs to allow the frontend to access it by adding the Access-Control-Allow-Origin in the response header. Something like this? 

UI: https://domain-a.com
ENGINE: https://domain-b.com

#!/usr/bin/env python3
from http.server import HTTPServer, SimpleHTTPRequestHandler
import sys

class CORSRequestHandler (SimpleHTTPRequestHandler):
    def end_headers (self):
        self.send_header('Access-Control-Allow-Origin', 'https://domain-a.com') 
        SimpleHTTPRequestHandler.end_headers(self)

Btw, I am not a developer, so I would not know how to implement this in the current solution. Otherwise, I would gladly contribute. :)

**Reply by DCMattyG** [reply]
I would say this is going to be difficult to troubleshoot without recreating this environment and testing on my end. There's a number of complexities that would be helpful to see for myself as I'd likely have to do some debugging.

Right now, we have a mechanism built-in to tell the Engine and UI that their URL's differ....

On the Engine container, you would set the "IPAM_UI_URL" variable, and on the UI container, you would set the "REACT_APP_IPAM_ENGINE_URL" variable. Both must be set.

Side note, is there a reason you are trying to use Azure Container Apps here instead of Azure App Services? Given that (to my knowledge) we don't have any other customers running this in ACA, I was just curious why this was your target path as opposed to something else many many people are using (hope you don't mind me asking).

More than happy to continue troubleshooting with you, but if you have any template or some way for me to recreate a "similar" ACA environment to what you're working with so that I can troubleshoot alongside you, that would help a lot.

As always, if you think this might be better served with a 1:1 Teams meeting, please don't hesitate to drop me an email at Matthew.Garrett@microsoft.com, and I'll do my best to set some time up for us to chat live.

**Reply by cata008** [reply]
Hi @DCMattyG, yes I concur, it will be difficult to troubleshoot it in this way. Unfortunately, at this stage I do not have any deployment template that I could share. I will drop you an email and maybe we can chat live :)

I have set both URLs "IPAM_UI_URL" and "REACT_APP_IPAM_ENGINE_URL". Still getting the CORS policy error. And while looking closer at the engine code, I can see that it is being implemented..

origins = [
    "http://localhost:3000"
]

if os.environ.get('WEBSITE_HOSTNAME'):
    origins.append("https://" + os.environ.get('WEBSITE_HOSTNAME'))

if os.environ.get('IPAM_UI_URL'):
    ui_url = urlparse(os.environ.get('IPAM_UI_URL'))

    if (ui_url.scheme and ui_url.netloc):
        origins.append(ui_url.scheme + "://" + ui_url.netloc)

app.add_middleware(
    CORSMiddleware,
    allow_origins = origins,
    allow_credentials = True,
    allow_methods = ["*"],
    allow_headers = ["*"],
)
I have deployed the solution using Azure Web Apps instead, and I instantly get pass this point.
<img width="311" alt="image" src="https://user-images.githubusercontent.com/37331541/237042148-1d13487a-c56d-4f1f-b058-49f61f6c0d7f.png">

Somehow, the allow credentials and origin headers are not returned when using a container app (see below). I even tried the CORS (preview) feature on the engine container, but it didn't make a difference.
<img width="364" alt="image" src="https://user-images.githubusercontent.com/37331541/237042955-f41f5ef3-a40a-4fdf-a4da-9dfbf71512cd.png">
 
Regarding your question about why I am using container apps, it was more me wanting to explore this option as in my ignorance I thought it should be pretty straightforward to implement it since we're working with containerized apps. :)

**Reply by DCMattyG** [reply]
Completely understood now @cata008, and no worries about trying out ACA, it would appear at first glance that it would be an ideal solution for Azure IPAM given it's container native approach. I do know that ACA has some nuances as to how the containers communicate internally, which differs from the Docker mechanism that Azure App Services utilizes.

I'll respond to your email and schedule some time for us to work together, and in the meantime I'll test ACA on my end as well to see if I can figure anything obvious out ahead of time.

Looking forward to solving this one with you!

**Reply by cata008** [reply]
@DCMattyG Thanks so much!

**Reply by DCMattyG** [reply]
Just to update anyone following this thread, the issue was related to some code bugs that have since been patched and pushed out to our hosted containers. Apologies for any inconvenience!
